src/Common/PipeWriterStream.cs

```diff
@@ -69,6 +69,11 @@ public override ValueTask WriteAsync(ReadOnlyMemory<byte> source, CancellationTo
 
         private ValueTask WriteCoreAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken = default)
         {
+            if (cancellationToken.CanBeCanceled)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Shouldn't this return a cancelled `ValueTask` instead of throw because there is no `async` keyword on the method?

Reply:
Why?

Reply:
@stephentoub are there any hard and fast rules here?

Reply:
The guidance is for the cancellation exception to come via the returned task rather than propagating out synchronously.  That generally goes for any exception other than usage exceptions like ArgumentException.  It doesn't matter at all if you're just awaiting, e.g.
```C#
await WriteAsync(...);
```
as then the caller can't tell the difference, but in other situations it does matter, such as if you're using combinators, using ContinueWith or some other method that operates on the returned task, storing the task and using it for later, etc.

Reply:
So basically don't use cancellationToken.ThrowIfCancellationRequested() unless the method is async.